erpnext/erpnext_integrations/doctype/leaflink_settings/leaflink_settings.py

```diff
@@ -0,0 +1,187 @@
+	"""
+	Sync an item in leaflink by hitting the Sync with LeafLink Button.
+	"""
+	pass
+
+
+@frappe.whitelist()
+def sync_customer_with_leaflink(customer_name):
+	"""
+	Sync a customer with leaflink by hitting Sync with Customer Button.
+	"""
+	leaflink_settings = frappe.get_cached_doc("LeafLink Settings")
+	header = {
+		leaflink_settings.get("api_key"): leaflink_settings.get("api_value")
+	}
+
+	#validate if leaflink is enabled
+	if not leaflink_settings.get("enable_leaflink"):
+		frappe.throw("Please enable LeafLink in LeafLink Settings.")
+
+	#check if customer exists on leaflink
+	try:
+		customers_url = leaflink_settings.get("customers_url")
+	except Exception as e:
+		frappe.throw("An Exception: {e} occured while trying to sync customer on LeafLink. Please try again.".format(e))
+
+	customers_on_leaflink = requests.get(customers_url, headers=header, params={"name": customer_name}).json()
+
+	#create customer record on leaflink if it does not exist
+	if not customers_on_leaflink.get("count"):
+
+		#create new customer data
+		new_customer = {
+			"name": customer_name,
+			"owner": leaflink_settings.get("leaflink_id")
+	 	}
+
+		#post request to create customer in leaflink
+		try:
+			post_response = requests.post(customers_url, headers=header, data=new_customer)
+			if post_response.status_code == 201:
+				frappe.msgprint("Customer successfully created on LeafLink!")
+		except Exception as e:
+			frappe.throw("An Exception: {e} occured while trying to sync customer on LeafLink. Please try again.".format(e))
+
+	else:
+		frappe.msgprint("Customer already exists on LeafLink!")
+
+
+@frappe.whitelist(allow_guest=True)
+def sales_order_from_leaflink(**args):
+
+	#create integration request
+	integration_request = frappe.new_doc("Integration Request")
+	integration_request.update({
+		"integration_type": "Host",
+		"integration_request_service": "LeafLink",
+		"status": "Queued",
+		"data":  json.dumps(args, default=json_handler),
+		"reference_doctype": "Sales Order"
+	})
+	integration_request.save(ignore_permissions=True)
+
+	#curate data from settings for post request
+	leaflink_settings = frappe.get_cached_doc("LeafLink Settings")
+	header = {
+		leaflink_settings.get("api_key"): leaflink_settings.get("api_value")
+	}
+
+	#ensure leaflink is enabled before proceeding further.
+	if not leaflink_settings.get("enable_leaflink"):
+		integration_request.update({
+			"status": "Failed",
+			"error": "Leaflink is not enabled."
+		})
+		integration_request.save(ignore_permissions=True)
+		return
+	sales_order = args.get("data")
+
+	#validate incoming keys
+	key = bytes(leaflink_settings.get("incoming_api_authentication_key"), 'utf-8')
+	incoming_signature = frappe.local.request.headers['LL-Signature'].encode("utf-8")
+	request_body = frappe.local.request.data
+	computed_signature = base64.b64encode(hmac.new(key, request_body, digestmod=hashlib.sha256).digest())
+	success = computed_signature == incoming_signature
```

Comment:
- This should be the first step that should be executed in the function
- Move this to a different function `authenticate_request` and then call it before leaflink request is executed

Reply:
The reason why its a not the first item, is because I want a log any requests being made to this endpoint. Hence, the integration request comes first, and then the authentication of the request. I'll move it to a separate function though